Refactor and improve docs

plaguss opened the pull request:
## Refactor and improve docs.

Closes #120

This PR is a refactor of the docs settings, including new sections, ased on the ideas from [diátaxis](https://diataxis.fr/). 
We now have a Learn section that contains tutorials and how-to guides (work in progress), and a technical reference that contains both concept guides and the API reference.

alvarobartt commented:
Hi here @plaguss, nice idea! I like the approach, feel free to open the PR for review once you're happy with it / is more mature 👍🏻

plaguss commented:
> @plaguss I also like the structure, and happy to help and split it between you and me. I'm still learning about distilabel, so I might need a little guidance, but eager to help.

we are together in that 😉

dvsrepo commented:
Hi @plaguss I like the structure too and the ref you make to diataxis. The only thing is that I feel the proposal of using the current examples as the main topics for the user guides is not very aligned. They are very quick examples and not so meaningful. The line between those and the tutorial is not so clean.

Things we need to document well:

Pipeline and pipeline

Tasks: what are the current tasks, their parameters, how to use these parameters 

For preference a description of the different options (UF, etc.), parameters, etc.

For text generation, how to use prompt templates, parameters, and how to use principles.

LLM: what are the current LLMs, their parameters, etc. Here for example we should help users to choose and option, to explain how llamacpp is used (which models?), how to setup vLLM with multiple GPUs (cc @gabrielmbmb )

Happy to do a call when I'm back from my holidays but I'd recommend keeping a mindset of adding parts describing key things in the most direct, descriptive, but simple language.

plaguss commented on docs/api/llms.md:
```diff
@@ -0,0 +1,17 @@
+# LLMs
+
+Already familiar with the `LLM` component? Otherwise you can take a look at the [concepts](../concepts.md) and come back later.
```
🙌

plaguss commented on docs/technical-reference/llms.md:
```diff
@@ -0,0 +1,152 @@
+
+The [`LLM`][distilabel.llm.base.LLM] class encapsulates the functionality for interacting with a language model. It distinguishes between *task* specifications and configurable arguments that influence the LLM's behavior. For illustration purposes, we employ the `TextGenerationTask` in this section and guide readers to the dedicated [`Tasks`](../technical-reference/tasks.md) section for comprehen ...
+
+To delineate their behavior, we have access to a series of arguments specific for each of them, but first let's see the general ones, and the `generate` method.
```
Regarding the distinction between parameter/argument, the first refers to the variable in a method or function definition while the second refers to the actual value that is passed to a function when it is called, so I think we should go with parameter here. For the `generate` I also think it's better to have a specific subsection

dvsrepo replied:
Sounds great 👍

plaguss commented:
I reviewed the comments and move all the code snippets to its corresponding folder. Had to add some `noqa` comments to avoid ruff in those specific cases, I think they aren't so annoying, but let me know

dvsrepo commented:
@plaguss can you replace the alembic svg with this:

![alembic](https://github.com/argilla-io/distilabel/assets/1107111/899eca5b-cee5-4242-87ea-5dd1ed732281)

dvsrepo commented:
good job, this is a very good version.

My only comment is that I think the noqa comments if displayed in the docs can be annoying and out of place (never seen this in other docs), don't we have a general way to exclude those code snippets from the qa analysis?

plaguss commented:
> good job, this is a very good version.
> 
> My only comment is that I think the noqa comments if displayed in the docs can be annoying and out of place (never seen this in other docs), don't we have a general way to exclude those code snippets from the qa analysis?

Done! my bad, I just removed the `/docs` folder from the ruff formatter (should have done that the first time).
I added the new alembic.svg, but I left the previous one because it's not working for the favicon (I don't know why).

alvarobartt commented on mkdocs.yml:
```diff
@@ -70,9 +71,21 @@ plugins:
       nav_file: SUMMARY.md
   - section-index
   - mkdocstrings
+  - mknotebooks
```
Are we planning to add notebooks atm? If not I think we're safe to remove this plugin for the moment

plaguss replied:
Removed

alvarobartt commented on mkdocs.yml:
```diff
@@ -70,9 +71,21 @@ plugins:
-  - Code Reference: reference/
+  - Learn:
+    - learn/index.md
+    - Tutorials:
+      - learn/tutorials/index.md
+    - User Guides:
+      - learn/user-guides/index.md
```
To me the difference between these sub-sections is not super clear, maybe we could unify those under `Tutorials`, not sure about the differences though

alvarobartt commented on docs/assets/alembic.svg:
Could we rename this to `logo.svg`?

plaguss replied:
fine by me

alvarobartt commented on mkdocs.yml:
```diff
@@ -13,13 +13,13 @@ extra_css:
 theme:
   name: material
-  logo: assets/distilabel-icon.svg
+  logo: assets/alembic.svg
```
The rename to logo mentioned before should also be changed here if you agree on applying it, otherwise feel free to discard 👍🏻 

```suggestion
  logo: assets/logo.svg
```

alvarobartt commented on docs/technical-reference/llms.md:
```diff
@@ -0,0 +1,101 @@
+
+For the API reference visit [vLLM][distilabel.llm.vllm.vLLM].
+
+## Huggingface LLMs
```
IMO we can rename all the occurrences of HuggingFace as either `HuggingFace` or `Hugging Face`

```suggestion
## HuggingFace LLMs
```

alvarobartt commented on docs/technical-reference/llms.md:
```diff
@@ -0,0 +1,101 @@
+
+## Huggingface LLMs
+
+This section explains two different ways to use huggingface's models:
```
```suggestion
This section explains two different ways to use HuggingFace models:
```

alvarobartt commented on docs/technical-reference/llms.md:
```diff
@@ -0,0 +1,101 @@
+
+### Inference Endpoints
+
+Hugging Face provides a streamlined approach for deploying models through [inference endpoints](https://huggingface.co/inference-endpoints) on their infrastructure. Opt for this solution if your model is hosted on Hugging Face.
```
e.g. here we use `Hugging Face`, previously we were using `Huggingface` and sometimes `huggingface`, so IMO we can just align those to whatever feels better? I'd say `HuggingFace` but I think that the naming standard is not too clear 🤣 

```suggestion
HuggingFace provides a streamlined approach for deploying models through [inference endpoints](https://huggingface.co/inference-endpoints) on their infrastructure. Opt for this solution if your model is hosted on HuggingFace.
```

alvarobartt commented on docs/technical-reference/pipeline.md:
```diff
@@ -0,0 +1,107 @@
+
+The [Pipeline][distilabel.pipeline.Pipeline] class is a central component in `distilabel`, responsible for crafting datasets. It manages the generation of datasets and oversees the interaction between the generator and labeller `LLMs`.
+
+You create an instance of the [`Pipeline`][distilabel.pipeline.Pipeline] by providing a *generator* and an optional *labeller* [LLM][distilabel.llm.base.LLM]. Interactions with it are facilitated through its `generate` method. This method requires a [`dataset`](https://huggingface.co/docs/datasets/v2.15.0/en/package_reference/main_classes#datasets.Dataset), specifies the *num_generations* to dete ...
```
Maybe we can point to the `main` branch instead of a pinned version?

```suggestion
You create an instance of the [`Pipeline`][distilabel.pipeline.Pipeline] by providing a *generator* and an optional *labeller* [LLM][distilabel.llm.base.LLM]. Interactions with it are facilitated through its `generate` method. This method requires a [`dataset`](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset), specifies the *num_generations* to determine the number of examples to be created, and includes additional parameters for controlling the *batch_size* and managing the generation process.
```

alvarobartt commented on docs/technical-reference/tasks.md:
```diff
@@ -0,0 +1,159 @@
+
+#### UltraFeedbackTask
+
+The task specially designed to build the prompts following the UltraFeedback paper: [UltraFeedback: Boosting Language Models With High Quality Feedback](https://arxiv.org/pdf/2310.01377.pdf).
```
```suggestion
The task specifically designed to build the prompts following the format defined in the ["UltraFeedback: Boosting Language Models With High Quality Feedback"](https://arxiv.org/pdf/2310.01377.pdf) paper.
```

alvarobartt commented on docs/technical-reference/tasks.md:
```diff
@@ -0,0 +1,159 @@
+
+Preference datasets for Language Models (LLMs) are sets of information that show how people rank or prefer one thing over another in a straightforward and clear manner. These datasets help train language models to understand and generate content that aligns with user preferences, enhancing the model's ability to generate contextually relevant and preferred outputs.
+
+Contrary to the `TextGenerationTask`, the `PreferenceTask` is not intended for direct use. It implements the default methods `input_args_names` and `output_args_names`, but `generate_prompt` and `parse_output` are specific to each `PreferenceTask`. Examining the output_args_names reveals that the generation will encompass both the rating and the rationale that influenced that rating. `distilabel` ...
```
```suggestion
Contrary to the `TextGenerationTask`, the `PreferenceTask` is not intended for direct use. It implements the default methods `input_args_names` and `output_args_names`, but `generate_prompt` and `parse_output` are specific to each `PreferenceTask`. Examining the `output_args_names` reveals that the generation will encompass both the rating and the rationale that influenced that rating.
```

alvarobartt commented on docs/technical-reference/tasks.md:
````diff
@@ -0,0 +1,159 @@
+
+The following example uses a `LLM` to examinate the data for text quality criteria, which includes the different criteria from UltraFeedback (Correctness & Informativeness, Honesty & Uncertainty, Truthfulness & Hallucination and Instruction Following):
+
+```python
+--8<-- "docs/snippets/technical-reference/tasks/openai_for_text_quality.py"
+```
+
+- Helpfulness:
+
+The following example creates a UltraFeedback task to emphasize helpfulness, that is overall quality and correctness of the output:
+
+```python
+--8<-- "docs/snippets/technical-reference/tasks/openai_for_helpfulness.py"
+```
+
+- Truthfulness:
+
+The following example creates a UltraFeedback task to emphasize truthfulness and hallucination assessment:
+
+```python
+--8<-- "docs/snippets/technical-reference/tasks/openai_for_truthfulness.py"
+```
+
+- Honesty:
+
+The following example creates a UltraFeedback task to emphasize honesty and uncertainty expression assessment:
+
+```python
+--8<-- "docs/snippets/technical-reference/tasks/openai_for_honesty.py"
+```
+
+- Instruction Following:
+
+The following example creates a UltraFeedback task to emphasize the evaluation of alignment between output and intent:
+
+```python
+--8<-- "docs/snippets/technical-reference/tasks/openai_for_instruction_following.py"
+```
+
+For the API reference visit [UltraFeedbackTask][distilabel.tasks.preference.ultrafeedback.UltraFeedbackTask].
````
Maybe we can keep these collapsed using content tabs? See more at https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#linked-content-tabs

plaguss replied:
yeah:
![image](https://github.com/argilla-io/distilabel/assets/56895847/a74394d5-98e4-42dc-bcd3-c7f5852623f3)

alvarobartt reviewed (commented):
LGTM! Thanks for this awesome (and much needed) PR @plaguss 🎉 I just have some comments that you can find below, but great work so far, it's been really straight forward to go through the docs!

P.S. Also I'm unsure about the value of keeping the scripts in separate files and then injecting them in the Markdown files, could you elaborate on that decision? How's that better to keeping those within the Markdown files? Thanks in advance 👍🏻

plaguss commented:
> LGTM! Thanks for this awesome (and much needed) PR @plaguss 🎉 I just have some comments that you can find below, but great work so far, it's been really straight forward to go through the docs!
> 
> P.S. Also I'm unsure about the value of keeping the scripts in separate files and then injecting them in the Markdown files, could you elaborate on that decision? How's that better to keeping those within the Markdown files? Thanks in advance 👍🏻

Thank you!!

The decision for keeping the scripts in the snippets directory instead of the Markdown files because we already had it that way for a couple of examples. I would be useful if we reused the code snippets in different parts of the docs (which is not the case currently, and if that was the case we could do it for a single case specifically). In this case I think it's more a matter of preference, but I'm fine with any decision, I can place them directly in the markdown files and remove the code snippets.

alvarobartt commented:
> > LGTM! Thanks for this awesome (and much needed) PR @plaguss 🎉 I just have some comments that you can find below, but great work so far, it's been really straight forward to go through the docs!
> > P.S. Also I'm unsure about the value of keeping the scripts in separate files and then injecting them in the Markdown files, could you elaborate on that decision? How's that better to keeping those within the Markdown files? Thanks in advance 👍🏻
> 
> Thank you!!
> 
> The decision for keeping the scripts in the snippets directory instead of the Markdown files because we already had it that way for a couple of examples. I would be useful if we reused the code snippets in different parts of the docs (which is not the case currently, and if that was the case we could do it for a single case specifically). In this case I think it's more a matter of preference, but I'm fine with any decision, I can place them directly in the markdown files and remove the code snippets.

Thanks for the answer! I'm happy keeping those within the docs instead, but may be opinionated, so lets double-check with @gabrielmbmb and @dvsrepo to see what are their thoughts 👍🏻

gabrielmbmb commented:
Hi @plaguss! Looks good to me, I saw there are some scripts using in the docs that got the license header injected. I think we should skip adding the license in those scripts.

@alvarobartt regarding the decision on having the code snippets in different files, I think it's better to have them in a `.py` file because is better for the development of them (you get assistance from the IDE and they can be formatted using tools) and also you can quickly execute them to test they are working (for this to be true, they have to be fully working examples)
